extensions/security/deployment/src/main/java/io/quarkus/security/deployment/PermissionSecurityChecks.java

```diff
@@ -847,4 +972,110 @@ private static boolean isStringPermission(PermissionKey permissionKey) {
         }
     }
 
+    final class PermissionConverterGenerator {
+        private final BuildProducer<GeneratedClassBuildItem> generatedClassesProducer;
+        private final BuildProducer<ReflectiveClassBuildItem> reflectiveClassesProducer;
+        private final SecurityCheckRecorder recorder;
+        private final Map<String, PermissionMethodConverter> converterNameToInstance = new HashMap<>();
+        private final IndexView index;
+
+        private PermissionConverterGenerator(BuildProducer<GeneratedClassBuildItem> generatedClassesProducer,
+                BuildProducer<ReflectiveClassBuildItem> reflectiveClassesProducer, SecurityCheckRecorder recorder,
+                IndexView index) {
+            this.generatedClassesProducer = generatedClassesProducer;
+            this.reflectiveClassesProducer = reflectiveClassesProducer;
+            this.recorder = recorder;
+            this.index = index;
+        }
+
+        private Map<String, PermissionMethodConverter> getConverterNameToInstance() {
+            return converterNameToInstance.isEmpty() ? null : Map.copyOf(converterNameToInstance);
+        }
+
+        private String createConverter(String paramRemainder, MethodInfo securedMethod, int methodParamIdx) {
+            String[] nestedParams = paramRemainder.split("\\.");
+            var converterName = createConverterClassName(securedMethod, methodParamIdx, 0);
+            try (ClassCreator classCreator = ClassCreator.builder()
```

Comment:
This seems to be generating one class per level of access of an expression, so for example given an expression `foo.bar.gee` we'll have to generate a `PermissionMethodConverter` class for `foo`, then `bar`, then `gee`.

This is probably fine, but more generally, I do wonder if we're not shooting ourselves in the foot generating single-method classes instead of grouping all the converters in a single class and use method handles to limit the number of classes loaded in the JVM.

WDYT @mkouba @geoand @cescoffier ? I still have Quarkus Intro slides that say we work to reduce the number of classes loaded, and yet a lot of extensions produce single-method classes everywhere. Do we have a preferred policy for extension authors?

Reply:
I haven't looked at this use case specifically, but I personally try (more than just a token effort) to generate as few classes as possible for each use case.

That said, I don't think we ever had a policy on this

Reply:
Yes, I made my life simple. What is written above is true. I can fix this by caching classes based on data types. I am not sure if it is worth it, as it will add complexity.

Reply:
Depends really on how many classes we typically expect to be generated. If it's in the low 10s, it's probably not worth the effort.

Reply:
We discussed this in https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Generated.20classes and it appears that it's generally better to get rid of the single-method-interface, generate all these methods as static methods in the same utility class, each with the same signature and a generated name, and use them as method handles.

That is a bit what I did in https://github.com/quarkusio/quarkus/pull/42642 where I avoided generating single-method classes to invoke record constructors and used method handles.

I suppose this doesn't have to be done for this PR, we could come back to it later, or not and just do this for later PRs.

Reply:
There is one downside to this proposal. Previously it was possible to also access fields and methods from same package (and sometimes in the same class -> private) while public methods / fields are required. Anyway, I don't think it's an issue because this implementation also supports getters, so users can simply define getters or use records (so we invoke method for that field instead of that field if it is not public).

I have tried it and it works, tomorrow I need to proceed to other comments of yours. Thanks for the suggestion.

Reply:
Ah, well, I didn't think of access indeed. We can always generate accessors if required. Later.